Add new guide: Connect a frontend - Closes #490 (#492)
LiskHQ/lisk-docs: modules/ROOT/pages/guides/frontend.adoc

= Connect a frontend
Mona Bärenfänger <mona@lightcurve.io>
:toc:
:imagesdir: ../../assets/images
:url_pug: https://pugjs.org/
:url_angular: https://angular.io/
:url_react: https://reactjs.org/
:url_vue: https://vuejs.org/

:url_github_sdk: https://github.com/LiskHQ/lisk-sdk

This guide describes how to create a frontend or client app to a Lisk blockchain network.

NOTE: A **Lisk blockchain network** is a network of nodes that are built with the {url_github_sdk}[Lisk SDK].

== Choose a framework

Choose based on personal preference.

> Some popular frameworks for developing apps are for example:

Reviewer: ```suggestion
Some popular framework examples for developing apps are listed below:
```

│   │   ├── index.html
│   │   ├── logo192.png
│   │   ├── logo512.png
│   │   ├── manifest.json
│   │   └── robots.txt
│   ├── src
│   │   ├── App.css
│   │   ├── accounts.json
│   │   ├── api.js
│   │   ├── assets
│   │   ├── components
│   │   │   ├── Accounts.js
│   │   │   ├── App.js
│   │   │   ├── Blocks.js
│   │   │   ├── Faucet.js
│   │   │   ├── Hello.js
│   │   │   ├── HelloAccounts.js
│   │   │   ├── HelloTransactions.js
│   │   │   ├── NewAccount.js
│   │   │   ├── Transactions.js
│   │   │   └── Transfer.js
│   │   ├── index.css
│   │   ├── index.js
│   │   └── logo.svg
│   └── yarn.lock
└── transactions
    ├── hello_transaction.js
    ├── index.js
    └── package.json
....

=== Install dependencies

Clone the {url_github_sdk_examples}[lisk-sdk-examples on GitHub] and navigate into `hello_world` folder.

First make sure, you have all xref:{url_setup}[Lisk SDK dependencies] installed, and have a fresh database `lisk_dev`.

Next, install all dependencies for the `hello_world` app and the `react-client` application.

.Install dependencies
[source,bash]
----
git clone https://github.com/LiskHQ/lisk-sdk-examples.git
cd lisk-sdk-examples/hello_world
npm i
cd react-client

> npm i

Reviewer 2: Given the guide is using yarn to start the react app, I'd suggest using yarn here as well.
```suggestion
yarn install
cd react-client
yarn install
```

... or the other way - use npm in both

Author: I will change it to npm, as we use it everywhere else in the docs :)

> ----

Reviewer 2: To me, it would make sense to have these bash commands and the other bash commands for starting the app next to each other (maybe here, maybe at the end) and group it by (1) instal and run backend and (2) install and run frontend.

"instal and run backend" might even link to another guide if we have it...

                </form>
                {this.state.response.meta.status &&
                <div>
                    <pre>Transaction: {JSON.stringify(this.state.transaction, null, 2)}</pre>
                    <p>Response: {JSON.stringify(this.state.response, null, 2)}</p>
                </div>
                }
            </div>
        );
    }
}
export default Hello;
----

=== Get all accounts

How to display all acounts on a web page.

image:hello-accounts.png[]

.react-client/src/components/Accounts.js
[source,jsx]
----
import React, { Component } from 'react';
import { api } from '../api.js';

class Accounts extends Component {

    constructor(props) {
        super(props);

        this.state = { data: [] };
    }

    async componentDidMount() {
        let offset = 0;
        let accounts = [];
        const accountsArray = [];

        do {
            const retrievedAccounts = await api.accounts.get({ limit: 100, offset });
            accounts = retrievedAccounts.data;
            accountsArray.push(...accounts);

            if (accounts.length === 100) {
                offset += 100;
            }
        } while (accounts.length === 100);

>         this.setState({ data: accountsArray });

Reviewer 2: I get this loop is fine for a PoC, but might be worth to add some comment explaining it should not be used in production as it doesn't scale with a lot of data.

Author: I added warnings everywhere where the loop is used :)
What would be a good approach to make it scale better? Probably adding some kind of pagination right?

Reviewer 2: Pagination, or a button to "Load more", or infinite scroll...

    }
}
export default HelloTransactions;
----

=== Get all blocks

How to display all blocks on a web page.

image:hello-blocks.png[]

.react-client/src/components/Blocks.js
[source,jsx]
----
import React, { Component } from 'react';
import { api } from '../api.js';

class Blocks extends Component {

    constructor(props) {
        super(props);

        this.state = { data: [] };
    }

    async componentDidMount() {
        let offset = 0;
        let blocks = [];
        const blocksArray = [];

        do {
            const retrievedBlocks = await api.blocks.get({ limit: 100, offset });
            blocks = retrievedBlocks.data;
            blocksArray.push(...blocks);

            if (blocks.length === 100) {
                offset += 100;
            }
        } while (blocks.length === 100);

        this.setState({ data: blocksArray });
    }

    render() {
        return (
            <div>
                <h2>All Blocks</h2>
                <pre>{JSON.stringify(this.state.data, null, 2)}</pre>
            </div>
        );
    }
}
export default Blocks;
----

== Preview in Browser

First, start the node that provides the API for the frontend:

[source,bash]
----

> node index.js | npx bunyan - o short

Reviewer 2: Typo:
```suggestion
node index.js | npx bunyan -o short
```

Also, guide assumes that I installed all the non-npm dependencies like postgres, created some database, but doesn't say.

import HelloTransactions from './HelloTransactions';
import Blocks from './Blocks';

// The pages of this site are rendered dynamically
// in the browser (not server rendered).

export default function App() {
    return (
        <Router>
            <div>
                <ul>
                    <li><Link to="/">Home</Link></li>
                    <hr />
                    <h3> Interact </h3>
                    <li><Link to="/new-account">New Account</Link></li>
                    <li><Link to="/faucet">Faucet</Link></li>
                    <li><Link to="/send-transfer">Send tokens</Link></li>
                    <li><Link to="/send-hello">Send Hello</Link></li>
                    <hr />
                    <h3> Explore </h3>
                    <li><Link to="/accounts">Accounts</Link></li>
                    <li><Link to="/hello-accounts">Hello accounts</Link></li>
                    <li><Link to="/transactions">Transactions</Link></li>
                    <li><Link to="/hello-transactions">Hello transactions</Link></li>
                    <li><Link to="/blocks">Blocks</Link></li>
                </ul>

                <hr />

                {/*

>                   A <Switch> looks through all its children <Route>

Reviewer: ```suggestion
                  A <Switch> looks through all of its child <Route> elements, and renders the first one whose path matches the current URL. Use a <Switch> whenever you have multiple routes, but only require one of them to be rendered at a time.
```

Author: This code is part of the react app, we should not change it.

Reviewer: Ah ah..OK then leave it..

                        <SendHello />
                    </Route>
                    <Route path="/accounts">
                        <Accounts />
                    </Route>
                    <Route path="/hello-accounts">
                        <HelloAccounts />
                    </Route>
                    <Route path="/blocks">
                        <Blocks />
                    </Route>
                    <Route path="/transactions">
                        <Transactions />
                    </Route>
                    <Route path="/hello-transactions">
                        <HelloTransactions />
                    </Route>
                </Switch>
            </div>
        </Router>
    );
}

/* Functional components */

function Home() {
    return (
        <div>
            <h2>Hello Lisk!</h2>
            <p>A simple frontend for blockchain applications built with the Lisk SDK.</p>
        </div>
    );
}
----

The `Home` component, which is defined as function at the bottom of `App.js`, is rendered as the start page:

image:hello-index.png[]

[TIP]
====
It is both possible to define components as functions or as classes.

> Although classes give more flexibility, functional components can be convenient. For example, as less code is needed to write them.

Reviewer: ```suggestion
Although classes give more flexibility, functional components can be convenient. For example, as less code is required to write them.
```

For reference, see also {url_medium_functional_vs_class}[Functional vs Class-Components in React].
====

== API client

> It is recommended to define the API client in a single file, and import it where needed.

Reviewer: ```suggestion
It is recommended to define the API client in a single file, and import it where required.
```

import * as cryptography from '@liskhq/lisk-cryptography';
import { Mnemonic } from '@liskhq/lisk-passphrase';

const newCredentials = () => {
    const passphrase = Mnemonic.generateMnemonic();
    const keys = cryptography.getPrivateAndPublicKeyFromPassphrase(
        passphrase
    );
    const credentials = {
        address: cryptography.getAddressFromPublicKey(keys.publicKey),
        passphrase: passphrase,
        publicKey: keys.publicKey,
        privateKey: keys.privateKey
    };
    return credentials;
};

class NewAccount extends Component {

    constructor(props) {
        super(props);

        this.state = { credentials: newCredentials() };
    }

    render() {
        return (
            <div>
                <h2>Create new account</h2>
                <p>Refresh page to get new credentials.</p>
                <pre>{JSON.stringify(this.state.credentials, null, 2)}</pre>
            </div>
        );
    }
}
export default NewAccount;
----

=== Faucet

A faucet is usually connected to a well funded account in the network, that is used as the source to send funds to accounts in order to get started.
It is therefore primarily useful for Devnets or Testnets.

> In this example, we use the genesis account in the faucet.

Reviewer: ```suggestion
In this example, the genesis account is used in the faucet.
```

> To keep it simple, we do not set any restrictions for the maximum amount and the frequency of requesting new tokens.

Reviewer: ```suggestion
In order to simplify this no restrictions are set for the maximum amount and the frequency of requesting new tokens.
```

        });
    }

    render() {
        return (
            <div>
                <h2>Faucet</h2>
                <p>The faucet transfers tokens from the genesis account to another.</p>
                <form onSubmit={this.handleSubmit}>
                    <label>
                        Address:
                        <input type="text" id="address" name="address" onChange={this.handleChange} />
                    </label>
                    <label>
                        Amount (1 = 10^8 tokens):
                        <input type="text" id="amount" name="amount" onChange={this.handleChange} />
                    </label>
                    <input type="submit" value="Submit" />
                </form>
                {this.state.response.meta.status &&
                    <div>
                        <pre>Transaction: {JSON.stringify(this.state.transaction, null, 2)}</pre>
                        <p>Response: {JSON.stringify(this.state.response, null, 2)}</p>
                    </div>
                }
            </div>
        );
    }
}
export default Faucet;
----

=== Send default transaction

How to send a default transaction from a web page.

> In this example, we send a transfer transaction.

Reviewer: ```suggestion
In this example, a transfer transaction is sent.
```

                <form onSubmit={this.handleSubmit}>
                    <label>
                        Recipient:
                        <input type="text" id="address" name="address" onChange={this.handleChange} />
                    </label>
                    <label>
                        Amount (1 = 10^8 tokens):
                        <input type="text" id="amount" name="amount" onChange={this.handleChange} />
                    </label>
                    <label>
                        Passphrase:
                        <input type="text" id="passphrase" name="passphrase" onChange={this.handleChange} />
                    </label>
                    <input type="submit" value="Submit" />
                </form>
                {this.state.response.meta.status &&
                <div>
                    <pre>Transaction: {JSON.stringify(this.state.transaction, null, 2)}</pre>
                    <p>Response: {JSON.stringify(this.state.response, null, 2)}</p>
                </div>
                }
            </div>
        );
    }
}
export default Transfer;
----

=== Send custom transactions

How to send a custom transaction from a web page.

> In this example, we send the hello transaction.

Reviewer: ```suggestion
In this example, a hello transaction is sent.
```

    render() {
        return (
            <div>
                <h2>Hello</h2>
                <p>Send a Hello transaction.</p>
                <form onSubmit={this.handleSubmit}>
                    <label>
                        Hello message:
                        <input type="text" id="hello" name="hello" onChange={this.handleChange} />
                    </label>
                    <label>
                        Passphrase:
                        <input type="text" id="passphrase" name="passphrase" onChange={this.handleChange} />
                    </label>
                    <input type="submit" value="Submit" />
                </form>
                {this.state.response.meta.status &&
                <div>
                    <pre>Transaction: {JSON.stringify(this.state.transaction, null, 2)}</pre>
                    <p>Response: {JSON.stringify(this.state.response, null, 2)}</p>
                </div>
                }
            </div>
        );
    }
}
export default Hello;
----

=== Get all accounts

How to display all acounts on a web page.

image:hello-accounts.png[]

[CAUTION]
====
The loop here is implemented in a very basic way to keep it simple.

> This is ok for proof of concept application with not much data.

Reviewer: ```suggestion
This is adequate for proof of concept applications with a minimal amount of data.
```

> For production networks, implement a solution that scales with a big amount of data.

Reviewer: ```suggestion
However, for production networks, it would be necessary to implement a solution that has the ability to scale with a large amount of data.
```

        let offset = 0;
        let accounts = [];
        let accountsArray = [];

        do {
            const retrievedAccounts = await api.accounts.get({ limit: 100, offset });
            accounts = retrievedAccounts.data;
            accountsArray.push(...accounts);

            if (accounts.length === 100) {
                offset += 100;
            }
        } while (accounts.length === 100);

        let assetAccounts = [];
        for (var i = 0; i < accountsArray.length; i++) {
            let accountAsset = accountsArray[i].asset;
            if (accountAsset && Object.keys(accountAsset).indexOf("hello") > -1){
                assetAccounts.push(accountsArray[i]);
            }
        }

        this.setState({ data: assetAccounts });
    }

    render() {
        return (
            <div>
                <h2>All Hello accounts</h2>
                <pre>{JSON.stringify(this.state.data, null, 2)}</pre>
            </div>
        );
    }
}
export default HelloAccounts;
----

=== Get all transactions

How to display all transactions on a web page.

image:hello-transactions.png[]

[CAUTION]
====
The loop here is implemented in a very basic way to keep it simple.

> This is ok for proof of concept application with not much data.

Reviewer: ```suggestion
This is adequate for proof of concept applications with a minimal amount of data.
```

> For production networks, implement a solution that scales with a big amount of data.

Reviewer: ```suggestion
However, for production networks, it would be necessary to implement a solution that has the ability to scale with a large amount of data.
```

====

.react-client/src/components/Transactions.js
[source,jsx]
----
import React, { Component } from 'react';
import { api } from '../api.js';

class Transactions extends Component {

    constructor(props) {
        super(props);

        this.state = { data: [] };
    }

    async componentDidMount() {
        let offset = 0;
        let transactions = [];
        const transactionsArray = [];

        do {
            const retrievedTransactions = await api.transactions.get({ limit: 100, offset });
            transactions = retrievedTransactions.data;
            transactionsArray.push(...transactions);

            if (transactions.length === 100) {
                offset += 100;
            }
        } while (transactions.length === 100);

        this.setState({ data: transactionsArray });
    }

    render() {
        return (
            <div>
                <h2>All Transactions</h2>
                <pre>{JSON.stringify(this.state.data, null, 2)}</pre>
            </div>
        );
    }
}
export default Transactions;
----

=== Get transactions by type

How to display all transactions of a particular type on a web page.

> In this example, we filter for hello transactions.

Reviewer: ```suggestion
In this example, it is shown how to filter for hello transactions.
```

image:hello-hello_transactions.png[]

.react-client/src/components/HelloTransactions.js
[source,jsx]
----
import React, { Component } from 'react';
import { api } from '../api.js';
import {
    HelloTransaction,
} from 'lisk-hello-transactions';

class HelloTransactions extends Component {

    constructor(props) {
        super(props);

        this.state = { data: [] };
    }

    async componentDidMount() {
        const transactions  = await api.transactions.get({ type: HelloTransaction.TYPE });

        this.setState({ data: transactions });
    }

    render() {
        return (
            <div>
                <h2>All Hello Transactions</h2>
                <pre>{JSON.stringify(this.state.data, null, 2)}</pre>
            </div>
        );
    }
}
export default HelloTransactions;
----

=== Get all blocks

How to display all blocks on a web page.

image:hello-blocks.png[]

[CAUTION]
====
The loop here is implemented in a very basic way to keep it simple.

> This is ok for proof of concept application with not much data.

Reviewer: ```suggestion
This is adequate for proof of concept applications with a minimal amount of data.
```

> For production networks, implement a solution that scales with a big amount of data.

Reviewer: ```suggestion
However, for production networks, it would be necessary to implement a solution that has the ability to scale with a large amount of data.
```

                            div.navbar-end
                                a(href='/create').navbar-item Create Credentials
                                a(href='/faucet').navbar-item Faucet
                                a(href='/post-hello').navbar-item Send Hello
                                a(href='/transfer').navbar-item Send Transfer
                                a.divider-vertical
                                a(href='/accounts').navbar-item Accounts
                                a(href='/hello-accounts').navbar-item Hello Accounts
                                a(href='/blocks').navbar-item Blocks
                                a(href='/transactions').navbar-item Transactions
                                a(href='/hello-transactions').navbar-item Hello Tx

        .container
            block content
                .jumbotron
                    h1 Hello Lisk!
                    p.lead
                        | A basic frontend for blockchain application uilt with the Lisk SDK.
----

This will create the following start page for the app:

image:hello-index.png[]

=== JavaScript

Create a file that will contain the JavaScript code for the client.

> We call it `app.js` in this example.

Reviewer: ```suggestion
In this example it is called `app.js`.
```

=== Node project

Clone the [lisk-sdk-examples on GitHub] and navigate into `hello_world/client` folder.

Or create a new client app manually.

> Note, that this client app is relying on a node with a reachable API, so it might be needed to adjust `API_BASEURL`.

Reviewer: ```suggestion
Note, that this client app is relying on a node with a reachable API, so it may be necessary to adjust `API_BASEURL`.
```

> In order to process custom transactions, it is needed to require these transaction in the client app as well.

Reviewer: ```suggestion
In order to process custom transactions, it is also necessary to require these transactions in the client app as well.
```

Create an intuitive and beautiful design for the app.

There are 2 different ways to include CSS styles in the client:

NOTE: The "Hello Lisk" client app uses CDN and local CSS styles in combination

==== CDN

To start quickly, it is possible to include external CSS via a CDN(= Content Delivery Network).

<<index_pug, The CDN styles are required in `views/index.pug`>>.

==== CSS files

> Alternatively or additionally to the CDN, it is possible to require local CSS files.

Reviewer: ```suggestion
Alternatively or additionally to the CDN, it is possible to require the local CSS files.
```

    res.render('hello-accounts', { accounts: assetAccounts });
});
----

== Start the app

> To start the web server, run:

Reviewer: ```suggestion
To start the web server, execute the following command:
```

[source,bash]
----
node app.js
# Explorer app listening on port 3000!
----

Now call http://localhost:3000 to view the client app in the browser.

> With all of the above, you should now have a basic frontend for the Hello world application, that allows users to

Reviewer: ```suggestion
At this stage, you should now have a basic frontend for the Hello world application, that allows users to perform the following:
```

* create new accounts
* send tokens to accounts
* get tokens from a faucet
* send `Hello` transactions
* explore all blocks
* explore all transactions and `Hello` transactions
* explore all `Hello` transactions
* explore all accounts and `Hello` accounts
* explore all `Hello` accounts

> TIP: Use this client as template or reference for your own client applications, and adjust it to your needs.

Reviewer: ```suggestion
TIP: Use this client as template or reference for your own client applications, and adjust it to suit your requirements.
```
